File: app/src/main/java/com/moataz/afternoonhadeeth/ui/adapter/BooksAdapter.java
package com.moataz.afternoonhadeeth.ui.adapter;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.moataz.afternoonhadeeth.R;
import com.moataz.afternoonhadeeth.data.model.books.Books;
import com.moataz.afternoonhadeeth.data.model.books.BooksResponse;
import com.moataz.afternoonhadeeth.databinding.ListBooksBinding;
import com.moataz.afternoonhadeeth.ui.view.activity.DisplayBookActivity;

import java.util.Collections;

public class BooksAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private BooksResponse items = null;

    @SuppressLint("NotifyDataSetChanged")
    public void setBooksList(BooksResponse items) {
        this.items = items;
        notifyDataSetChanged();
        Collections.shuffle(items.getBooksList());
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new BooksViewHolder(
                DataBindingUtil.inflate(
                        LayoutInflater.from(parent.getContext()),
                        R.layout.list_books,
                        parent,
                        false
                )
        );
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        Books books = items.getBooksList().get(position);
        ((BooksViewHolder) holder).listBooksBinding.setBooksModel(books);
        ((BooksViewHolder) holder).setOnClick(books);
    }

    @Override
    public int getItemCount() {
        if (items == null) return 0;
        return items.getBooksList().size();
    }

    static class BooksViewHolder extends RecyclerView.ViewHolder {
        ListBooksBinding listBooksBinding;

        BooksViewHolder(@NonNull ListBooksBinding itemView) {
            super(itemView.getRoot());
            listBooksBinding = itemView;
            listBooksBinding.ratingBar.setRating((float) 5);
        }

        void setOnClick(Books books) {
            itemView.setOnClickListener(v -> {
                Intent intent = new Intent(itemView.getContext(), DisplayBookActivity.class);
                // send data to DisplayBookActivity
                intent.putExtra("bookTitle", books.getTitle());
                intent.putExtra("bookAuthor", books.getAuthor());
                intent.putExtra("bookDetails", books.getDetails());
                intent.putExtra("bookImage", books.getImageURL());
                intent.putExtra("bookUrl", books.getUrl());
                itemView.getContext().startActivity(intent);
            });
        }
    }
}

File: app/src/main/java/com/moataz/afternoonhadeeth/ui/books/adapter/BooksAdapter.kt
package com.moataz.afternoonhadeeth.ui.books.adapter

import android.annotation.SuppressLint
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.RecyclerView
import com.moataz.afternoonhadeeth.R
import com.moataz.afternoonhadeeth.data.model.books.Books
import com.moataz.afternoonhadeeth.data.model.books.BooksResponse
import com.moataz.afternoonhadeeth.databinding.ListBooksBinding
import com.moataz.afternoonhadeeth.ui.books.view.activity.DisplayBookActivity

class BooksAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var items: BooksResponse? = null

    @SuppressLint("NotifyDataSetChanged")
    fun setBooksList(items: BooksResponse) {
        this.items = items
        notifyDataSetChanged()
        items.booksList.shuffled()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return BooksViewHolder(
            DataBindingUtil.inflate(
                LayoutInflater.from(parent.context),
                R.layout.list_books,
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val books = items!!.booksList[position]
        (holder as BooksViewHolder).listBooksBinding.booksModel = books
        holder.setOnClick(books)
    }

    override fun getItemCount(): Int {
        return if (items == null) 0 else items!!.booksList.size
    }

    internal class BooksViewHolder(var listBooksBinding: ListBooksBinding) :
        RecyclerView.ViewHolder(
            listBooksBinding.root
        ) {
        fun setOnClick(books: Books) {
            itemView.setOnClickListener { v: View? ->
                val intent = Intent(itemView.context, DisplayBookActivity::class.java)
                // send data to DisplayBookActivity
                intent.putExtra("bookTitle", books.title)
                intent.putExtra("bookAuthor", books.author)
                intent.putExtra("bookDetails", books.details)
                intent.putExtra("bookImage", books.imageURL)
                intent.putExtra("bookUrl", books.url)
                itemView.context.startActivity(intent)
            }
        }

        init {
            listBooksBinding.ratingBar.rating = 5f
        }
    }
}